contracts/ronin/validator/CoinbaseExecution.sol

```diff
@@ -112,36 +111,43 @@ abstract contract CoinbaseExecution is
    * @inheritdoc ICoinbaseExecution
    */
   function wrapUpEpoch() external payable virtual override onlyCoinbase whenEpochEnding oncePerEpoch {
-    uint256 newPeriod = _computePeriod(block.timestamp);
-    bool periodEnding = _isPeriodEnding(newPeriod);
-
-    address[] memory currValidatorIds = getValidatorIds();
-    address[] memory revokedCandidateIds;
-    uint256 epoch = epochOf(block.number);
-    uint256 nextEpoch = epoch + 1;
-    uint256 lastPeriod = currentPeriod();
-
-    _syncFastFinalityReward(epoch, currValidatorIds);
-
-    if (periodEnding) {
-      (uint256 totalDelegatingReward, uint256[] memory delegatingRewards) =
-        _distributeRewardToTreasuriesAndCalculateTotalDelegatingReward(lastPeriod, currValidatorIds);
-      _settleAndTransferDelegatingRewards(lastPeriod, currValidatorIds, totalDelegatingReward, delegatingRewards);
-      _tryRecycleLockedFundsFromEmergencyExits();
-      _recycleDeprecatedRewards();
-
-      ISlashIndicator slashIndicatorContract = ISlashIndicator(getContract(ContractType.SLASH_INDICATOR));
-      slashIndicatorContract.execUpdateCreditScores(currValidatorIds, lastPeriod);
-      (currValidatorIds, revokedCandidateIds) = _syncValidatorSet(newPeriod);
-      if (revokedCandidateIds.length > 0) {
-        slashIndicatorContract.execResetCreditScores(revokedCandidateIds);
+    unchecked {
+      uint256 newPeriod = _computePeriod(block.timestamp);
+      bool periodEnding = _isPeriodEnding(newPeriod);
+
+      uint256 lastPeriod = currentPeriod();
+      uint256 epoch = epochOf(block.number);
+      uint256 nextEpoch = epoch + 1;
+      address[] memory currValidatorIds = getValidatorIds();
+
+      IRandomBeacon randomBeacon = IRandomBeacon(getContract(ContractType.RANDOM_BEACON));
+      randomBeacon.onWrapUpEpoch(lastPeriod, newPeriod);
```

Comment:
How about split into two different calls. One is only called at the ending of period for finalizing beacon. And the other one is called only at the beginning of period, to request the new seed. This will reduce internal calls, and bring the check from BeaconContract to this, however will increase the size for RoninValidatorSet.

Reply:
I prefer grouping logics in the same place for consistency

Reply:
I see no consistency here, because they actually are two different works underneath. The only common thing is they are both invoked in "wrap up epoch", but it really not, while one is at the end of period, and the other one is at each epoch.

Reply:
"both are invoked by wrapUpEpoch" seems not convincing enough though